### PR TITLE
`Adaptive learning`: Improve usability, performance and code structure of competency management

### PR DESCRIPTION
**Pull request by rstief**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
The competency management has multiple problems:
- The amount of buttons in the header (now 4, will be 5) is too large
- Each time something is deleted/created, the server is called 2x per competency
- The `competency-management.component` file fulfills multiple roles and is too large

### Description
Client:
- Reduced text on buttons, changed import button to be a dropdown, since there will be up to three ways to import competencies
- Improved table layout (see screenshots for details)
- Adjusted multiple parts of the code to reduce complexity and function calls in the template
- Removed API calls on every change, instead changing the loaded data
- Extracted the competency relation graph to its own component

Server:
- Adjusted the competency relation endpoints to reduce API calls and to be more in-line with endpoint naming conventions

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 course with competencies, e.g. [Raphael Test Course](https://artemis-test1.artemis.cit.tum.de/course-management/16/competency-management) (or create new ones)

1. Log in to Artemis
2. Navigate to Course Management -> Competencies
4. Test creating and deleting (click the relation text) all kinds of competency relations
5. Test deleting competencies, see that the graph gets updated
6. Test deleting pre-requisites (the alert message should reference prerequisites, not competencies) 
7. Test that the tables still look good when re-sizing the screen 

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [competency-management.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/competencies/competency-management/competency-management.component.ts.html) | 92.5% | ✅  |
| [competency-relation-graph.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/competencies/competency-management/competency-relation-graph.component.ts.html) | 93.93% | ✅  |
| [competency.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/competencies/competency.service.ts.html) | 94.44% | ✅  |
| [competency.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/competency.model.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [CompetencyRelationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.competency/CompetencyRelationService.html) | 100% | ✅ |
| [CompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6150/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CompetencyResource.html) | 98% | ✅  |

#### Screenshots
Notable UI changes:
- dropdown button for import
- improved table layout for large screns (especially more space for competency description)
- improved table layout for small screns (remove mastered students **and recommended until**, improved button layout)
- New title for competency/prerequisite table
- Message if table is empty

**New layout:**
image

**Old layout**
image

**New layout (small):**
image

**Old layout (small):**
image

## Summary by CodeRabbit

- **New Features**
	- Introduced a new competency relation graph component for visual management of competency relations.
	- Added functionality for creating, editing, and deleting competency relations through both graphical interface and management components.
	- Implemented new `CompetencyRelationType` enum for better classification of competency relations.

- **Enhancements**
	- Enhanced competency management layout, improving user interaction for competency creation, editing, and deletion.
	- Improved authorization checks and error handling in competency relations management.
	- Refactored competency and relation management logic for clarity and maintainability.

- **Bug Fixes**
	- Fixed issues related to competency relation creation and deletion, ensuring a smoother user experience.

- **Refactor**
	- Significant code refactorings and optimizations in competency management components and services.
	- Renamed `CompetencyTableComponent` to `ImportCompetenciesTableComponent` for better clarity.

- **Style**
	- Added styling for the new competency relation graph component, enhancing visual presentation.

- **Tests**
	- Updated and added new tests for competency management and relation components to ensure reliability.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyRelationService.java [hunk 1]
```diff
@@ -57,27 +58,30 @@ public CompetencyRelation getCompetencyRelation(Competency tailCompetency, Compe
...
     }
+
+    /**
+     * Converts a CompetencyRelation to a CompetencyRelationDTO
+     *
+     * @param relation the relation
+     * @return the CompetencyRelationDTO
+     */
+    public CompetencyRelationDTO relationToDTO(CompetencyRelation relation) {
+        return new CompetencyRelationDTO(relation.getId(), relation.getTailCompetency().getId(), relation.getHeadCompetency().getId(), relation.getType());
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyRelationService.java:85]: Could we move this method to the DTO? (`CompetencyRelationDTO.of(CompetencyRelation relation)`)

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/test/javascript/spec/component/competencies/competency-management/competency-relation-graph-stub.component.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { Component, EventEmitter, Output, input } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (rstief) [reply src/test/javascript/spec/component/competencies/competency-management/competency-relation-graph-stub.component.ts:1]: this is incorrect. it uses input signals which are lowercase

File: src/main/webapp/app/course/competencies/competency-management/competency-relation-graph.component.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        }
+        if (this.containsCircularRelation()) {
+            this.relationError = CompetencyRelationError.CIRCULAR;
+            return;
+        }
+        this.relationError = undefined;
+    }
+
+    /**
+     *
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/course/competencies/competency-management/competency-relation-graph.component.ts:106]: missing doc string :)

**Review by reschandreas — APPROVED** [review]
overall code lgtm, thanks for refactoring this!

**Review by laurenzfb — COMMENTED** [review]
code

**Review by laurenzfb — APPROVED** [review]
code

**Comment by laurenzfb** [comment]
image
Light mode issues :(

**Review by milljoniaer — APPROVED** [review]
tested on testing session, works as expected

**Review by laurenzfb — APPROVED** [review]
lgtm

**Review by rabeatwork — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by Predixx — APPROVED** [review]
tested on ts6. lgtm!

**Review by vinceclifford — APPROVED** [review]
Tested manually on TS6 with Liam Berger. Works as expected!

**Pull request merged**